### DISCUSSION TITLE
Azimuth/Elevation to RA/Dec conversion functions?

### DISCUSSION THREAD

**Discussion by MrForum** [Stellarium/stellarium#2115] [Q&A] [answered]
Hello - I have Azimuth/Elevation coordinates in degrees and am wondering how to convert them to Right Ascension / Declination coordinates (for the current observer location and time) using utility functions (StelCore / StelUtils) from a Stellarium plugin.  

I see the PointerCoordinates plugin doing something similar starting from a mouse pointer location, so I imagine there's a built-in way to do it, but that plugin starts from core->getMouseJ2000Pos().

Thanks.

**Comment by gzotti** [comment]
These functions are not in the scripting API. You can implement them yourself, see e.g. https://en.wikipedia.org/wiki/Astronomical_coordinate_systems#Converting_coordinates.

To get sidereal time, use `core.getObjectInfo(objectName)` for any object ("Sun" should work) and retrieve the meanSidTm or appSidTm values.

Ah sorry, plugin. Use `StelCore::altAzToEquinoxEqu(const Vec3d& v, RefractionMode refMode)`. Before calling, set time to the time you need.

**Reply by MrForum** [reply]
@gzotti Thanks for the help.  For inputs, how would I go from azimuth/elevation in degrees to a properly constructed Vec3d?  And on the output Vec3d from altAzToEquinoxEqu, would I use StelUtils::rectToSher to get RA and Dec in radians?

**Reply by alex-w** [reply]
See example:
```
double longitude = 25. // degrees
double latitude = -5. // degrees
Vec3d XYZ;
// convert degrees into radians and compute vector
StelUtils::spheToRect(longitude * M_PI_180, latitude * M_PI_180, XYZ);
```

**Reply by gzotti** [reply]
Please read essentials on coordinate conversion at https://en.wikipedia.org/wiki/Spherical_coordinate_system. 
The Vec3d should be normalized, i.e. have length 1. 
Yes, rectToSphe(double *lng, double *lat, const Vec3d& v) converts v to the respective longitude and latitude coordinates. Note however that azimuth is counted in the opposite direction, so you may need to compute az=2pi - lng. (or az=pi-lng, depending on your preference). A bit of experimentation will be required to work that out.

**Reply by MrForum** [reply]
@gzotti 
So it sounds like this approach should work assuming I can get step 1 correct: 

1) Convert az/el degrees to unit vector (azElVec)
2) Vec3d eqVec = StelCore::altAzToEquinoxEqu(azElVec)
3) double raRads, decRads; StelUtils::rectToSphe(&raRads,&decRads,eqVec)

I will try this to implement step 1: https://math.stackexchange.com/questions/1150232/finding-the-unit-direction-vector-given-azimuth-and-elevation

Let me know if anything looks incorrect, thank you both.

**Reply by gzotti** [reply]
Yes, looks OK. Take care about degrees and radians, and as said the direction of counting longitude/azimuth. Good luck!

**Reply by MrForum** [reply]
This approach hasn't produced correct results so far, or even results that are consistently off by some amount of rotation.  The attached output is an example where a marker placed at az = 84.6 deg, el = 27.6deg should have a RA/Dec of 1h16m/20d56m (shown via Pointer Coordinates).  Results of my function and all intermediate calculations are shown at the bottom of the image.  

```
RADecPoint MyConnector::raDecTo(Alert alert)
{
	AzElPoint azEl = azElTo(alert); 
	qDebug() << "input azimuth (deg): " << azEl.azDeg << ", input elevation (deg): " << azEl.elDeg;

	double azRads = (azEl.azDeg * M_PI)/180.0;
	double elRads = (azEl.elDeg * M_PI)/180.0;
	qDebug() << "input azimuth (rad): " << azRads << ", input elevation (rad): " << elRads;
	
	Vec3d altAzVec(std::sin(azRads)*std::cos(elRads),std::cos(azRads)*std::cos(elRads),std::sin(elRads));
	qDebug() << "altAzVec: "<< altAzVec[0] << ", " << altAzVec[1] << ", " << altAzVec[2];

	Vec3d eqVec = core->altAzToEquinoxEqu(altAzVec);
	qDebug() << "eqVec: "<< eqVec[0] << ", " << eqVec[1] << ", " << eqVec[2];

	double raRads, decRads;
	StelUtils::rectToSphe(&raRads,&decRads,eqVec);
	qDebug() << "RA rads: " << raRads << ", decRads: " << decRads; 
	
	RADecPoint retVal;
	retVal.raDeg = raRads * 180/M_PI; 
	retVal.decDeg = decRads * 180/M_PI;
	qDebug() << "RA deg: " << std::to_string(retVal.raDeg).c_str() << ", deg: " << std::to_string(retVal.decDeg).c_str(); 
	return retVal; 
}

RADecPoint raDec = raDecTo(alert); 
QString ra = StelUtils::radToHmsStr(raDec.raDeg*M_PI/180.0, true);
QString dec = StelUtils::radToDmsStr(raDec.decDeg*M_PI/180.0, true);
qDebug() << "RA: " << ra << ", Dec: " << dec;
```

![image](https://user-images.githubusercontent.com/83614894/146989161-7976bab1-8515-48cf-b776-f91a06287ad3.png)

**Reply by alex-w** [reply]
What is main target for markers in your case? Just mark some place on the sky or not? In some cases CustomObjects will be more powerful.

**Comment by MrForum** [comment, marked as answer]
@alex-w yes it looks like if I add a custom object by azimuth/elevation rather than a marker this is all handled for me; I can even call the various coordinate functions on the object to easily convert between systems. Thanks for the suggestion.
